packages/subgraph/src/utils/tokens/erc1155.ts

```diff
@@ -0,0 +1,345 @@
+  getERC1155TransferId,
+  getTokenIdBalanceId
+} from './common';
+
+export function supportsERC1155(token: Address): bool {
+  // Double check that it's ERC1155 by calling supportsInterface checks.
+  let erc1155 = ERC1155.bind(token);
+  let introspection_ERC165 = supportsInterface(erc1155, ERC165_MAGIC_VALUE); // ERC165
+  let introspection_ERC1155 = supportsInterface(erc1155, ERC1155_MAGIC_VALUE); // ERC1155
+  let introspection_00000000 = supportsInterface(erc1155, '00000000', false);
+  return (
+    introspection_ERC165 && introspection_ERC1155 && introspection_00000000
+  );
+}
+
+export function fetchERC1155(address: Address): ERC1155Contract | null {
+  // Try load entry
+  let contract = ERC1155Contract.load(address.toHexString());
+  if (contract != null) {
+    return contract;
+  }
+
+  // Detect using ERC165
+  if (!supportsERC1155(address)) {
+    return null;
+  }
+  contract = new ERC1155Contract(address.toHexString());
+  contract.save();
+
+  return contract;
+}
+
+export function updateERC1155Balance(
+  daoId: string,
+  token: string,
+  tokenId: BigInt,
+  amount: BigInt,
+  timestamp: BigInt,
+  type: TransferType
+): void {
+  // generate unique balance id
+  let balanceId = getBalanceId(daoId, token);
+  // load balance
+  let erc1155Balance = ERC1155Balance.load(balanceId);
+
+  if (!erc1155Balance) {
+    // if not exists create it
+    erc1155Balance = new ERC1155Balance(balanceId);
+    erc1155Balance.dao = daoId;
+    erc1155Balance.token = token;
+    let erc1155 = ERC1155.bind(Address.fromString(token));
+    let metadataUri = erc1155.uri(tokenId);
+    erc1155Balance.metadataUri = metadataUri;
+  }
+  // update lastUpdated
+  erc1155Balance.lastUpdated = timestamp;
+  // generate unique tokenIdBalanceId
+  let tokenIdBalanceId = getTokenIdBalanceId(daoId, token, tokenId);
+  // load tokenIdbalnce
+  let erc1155TokenIdBalance = ERC1155TokenIdBalance.load(tokenIdBalanceId);
+  // if not exists create it
+  if (!erc1155TokenIdBalance) {
+    erc1155TokenIdBalance = new ERC1155TokenIdBalance(tokenIdBalanceId);
+    erc1155TokenIdBalance.amount = amount;
+    erc1155TokenIdBalance.tokenId = tokenId;
+    erc1155TokenIdBalance.balance = balanceId;
+  } else {
+    // if exists update it
+    if (type == TransferType.Deposit) {
+      erc1155TokenIdBalance.amount = erc1155TokenIdBalance.amount.plus(amount);
+    } else {
+      erc1155TokenIdBalance.amount = erc1155TokenIdBalance.amount.minus(amount);
+    }
+  }
+  // update lastUpdated
+  erc1155TokenIdBalance.lastUpdated = timestamp;
+  // save
+  erc1155Balance.save();
+  erc1155TokenIdBalance.save();
+}
+
+export function handleERC1155Received(
+  token: Address,
+  dao: Address,
+  data: Bytes,
+  event: ethereum.Event
+): void {
+  let contract = fetchERC1155(token);
+  if (!contract) {
+    return;
+  }
+  let calldata = data.toHexString().slice(10);
+  calldata = DECODE_OFFSET.concat(calldata);
+  let decodeABI = '(address,address,uint256,uint256,bytes)';
```

Comment:
@Rekard0 I'm unsure about the DECODE_OFFSET, can you check that is correct

Reply:
it is ok go a head